packages/truffle-debugger/lib/evm/selectors/index.js

```diff
@@ -481,12 +472,46 @@ const evm = createSelectorTree({
 
       /**
        * evm.current.step.isExceptionalHalting
-       *
        */
       isExceptionalHalting: createLeaf(
-        ["./isHalting", "/current/state/depth", "/next/state/depth"],
-        (halting, currentDepth, nextDepth) =>
-          nextDepth < currentDepth && !halting
+        [
+          "./isHalting",
+          "/current/state/depth",
+          "/next/state/depth",
+          "./returnStatus"
+        ],
+        (halting, currentDepth, nextDepth, status) =>
+          halting
+            ? !status //if deliberately halting, check the return status
+            : nextDepth < currentDepth //if not on a deliberate halt, any halt
+        //is an exceptional halt
+      ),
+
+      /**
+       * evm.current.step.returnStatus
+       * checks the return status of the *current* halting instruction (for
+       * normal halts only)
+       * (returns a boolean -- true for success, false for failure)
+       */
+      returnStatus: createLeaf(
+        [
+          "./isHalting",
+          "/next/state",
+          trace.stepsRemaining,
+          "/transaction/status"
+        ],
+        (matches, { stack }, remaining, finalStatus) => {
```

Comment:
I confirmed the related issue in codeclimate for this: I think we can get around this with `s/matches/isHalting`, and it'll also be a bit more clear at a glance what "to match" means.

Reply:
Do you think this should also be changed elsewhere?